parser/charset/encoding.go

```diff
@@ -99,6 +99,36 @@ func (e *Encoding) Encode(dest, src []byte) ([]byte, error) {
 	return e.transform(e.enc.NewEncoder(), dest, src, false)
 }
 
+// EncodeInternal convert bytes from utf-8 charset to a specific charset, we actually do not do the real convert, just find the inconvertible character and use ? replace.
+// The code below is equivalent to
+//		expr, _ := e.Encode(dest, src)
+//		ret, _ := e.Decode(nil, expr)
+//		return ret
+func (e *Encoding) EncodeInternal(dest, src []byte) []byte {
+	if !e.Enabled() {
+		return src
+	}
+	if dest == nil {
+		dest = make([]byte, 0, len(src))
+	}
+	var srcOffset int
+
+	buf := make([]byte, 4)
```

Comment:
```suggestion
	var buf [4]byte
```

Reply:
We can avoid the allocation from heap.

Reply:
no, we cannot, [4]byte is different from []byte.